core/src/main/scala/za/co/absa/spline/harvester/plugin/registry/AutoDiscoveryPluginRegistry.scala

```diff
@@ -0,0 +1,105 @@
+import javax.annotation.Priority
+import org.apache.commons.lang.ClassUtils.{getAllInterfaces, getAllSuperclasses}
+import org.apache.spark.internal.Logging
+import za.co.absa.commons.lang.ARM
+import za.co.absa.spline.harvester.plugin.Plugin
+import za.co.absa.spline.harvester.plugin.Plugin.Precedence
+import za.co.absa.spline.harvester.plugin.registry.AutoDiscoveryPluginRegistry.{PluginClasses, getOnlyOrThrow}
+
+import scala.collection.JavaConverters._
+import scala.reflect.ClassTag
+import scala.util.Try
+import scala.util.control.NonFatal
+
+class AutoDiscoveryPluginRegistry(injectables: AnyRef*)
+  extends PluginRegistry
+    with Logging {
+
+  private val injectablesByType: Map[Class[_], Seq[_ <: AnyRef]] = {
+    val typedInjectables =
+      for {
+        o <- this +: injectables
+        c = o.getClass
+        t <- getAllSuperclasses(c).asScala ++ getAllInterfaces(c).asScala :+ c
+      } yield t.asInstanceOf[Class[_]] -> o
+    typedInjectables.groupBy(_._1).mapValues(_.map(_._2))
+  }
+
+  private val allPlugins: Seq[Plugin] =
+    for (pc <- PluginClasses) yield {
+      log.info(s"Loading plugin: $pc")
+      instantiatePlugin(pc)
+        .recover({ case NonFatal(e) => throw new RuntimeException(s"Plugin instantiation failure: $pc", e) })
+        .get
+    }
+
+  override def plugins[A: ClassTag]: Seq[Plugin with A] = {
+    val ct = implicitly[ClassTag[A]]
+    allPlugins.collect({ case p: Plugin with A if ct.runtimeClass.isInstance(p) => p })
+  }
+
+  private def instantiatePlugin(pluginClass: Class[_]): Try[Plugin] = Try {
+    val constrs = pluginClass.getConstructors
+    val constr = getOnlyOrThrow(constrs, s"Cannot instantiate plugin with multiple constructors: ${constrs.mkString(", ")}")
+    val args = constr.getParameterTypes.map(pt => {
+      val candidates = injectablesByType.getOrElse(pt, sys.error(s"Cannot bind $pt. No value found"))
+      getOnlyOrThrow(candidates, s"Ambiguous constructor parameter binding. Multiple values found for $pt: ${candidates.length}")
+    })
+    constr.newInstance(args: _*).asInstanceOf[Plugin]
+  }
+
+}
+
+object AutoDiscoveryPluginRegistry extends Logging {
+
+  private implicit object PluginClassOrdering extends Ordering[Class[Plugin]] {
+    private def priority(c: Class[_]): Int =
+      Option(c.getAnnotation(classOf[Priority]))
+        .map(_.value)
+        .getOrElse(Precedence.User)
+
+    override def compare(x: Class[Plugin], y: Class[Plugin]): Int =
+      priority(x) - priority(y)
+  }
+
+  private val PluginClasses: Seq[Class[Plugin]] = {
+    log.debug("Scanning for plugins")
+    val classGraph = new ClassGraph().enableClassInfo
+    for {
+      scanResult <- ARM.managed(classGraph.scan)
+      pluginClass <- scanResult
+        .getClassesImplementing(classOf[Plugin].getName)
+        .loadClasses
+        .asScala
+        .asInstanceOf[Seq[Class[Plugin]]]
+        .sorted
+    } yield {
+      log.debug(s"Found plugin: $pluginClass")
```

Comment:
For debugging, it would be nice to print priority of all loaded plugins

Reply:
done